### ISSUE TITLE
Member `operator new` and `operator delete` should reject explicit object parameters

### ISSUE THREAD

**Issue by Rajveer100** [llvm/llvm-project#82249]
These functions are members, but they are implicitly static members as in this [draft](https://eel.is/c++draft/class#free-3), which means they shouldn't be allowed to have explicit object parameters.

Snippet:

```C++
#include <new>
#include <memory>

struct S {
    void *operator new(this unsigned long);
    void operator delete(this void*);
    void operator delete(this S*, std::destroying_delete_t);
};
```

[GodBolt](https://godbolt.org/z/vG16GEzGM)

**Comment by shafik** [comment]
The fix for this may be similar to: https://github.com/llvm/llvm-project/pull/80327 

but where `err_explicit_object_parameter_nonmember` diagnostic is handled.

**Comment by shafik** [comment]
I used a slightly reduced version to test: https://godbolt.org/z/z7jd8nnh7

I am surprised only MSVC diagnoses this.

The fix would be to modify:

https://github.com/llvm/llvm-project/blob/54b014b3f76e1c0060bd129e1196b6c729cb30b0/clang/lib/Sema/SemaDeclCXX.cpp#L11398-L11402

as follows:

```cpp
if (D.getDeclSpec().getStorageClassSpec() == DeclSpec::SCS_static ||
      (D.getContext() == clang::DeclaratorContext::Member &&
          D.isStaticMember())) {
```

This change diagnoses and passes `check-clang` I am bit surprised we don't obtain more diagnostics though.

CC @cor3ntin

**Comment by shafik** [comment]
Apologies, when I first opened this issue there was not a PR against it and I just realized there is a PR now with the exact solution I was proposing.

**Comment by frederick-vs-ja** [comment]
This should already be resolved...

If `Resolves #82249` were written instead of `Resolves Issue #82249` in the decription of that PR, this issue would be automatically closed.
